src/basho_bench_worker.erl

```diff
@@ -334,3 +347,25 @@ rate_worker_run_loop(State, Lambda) ->
         ExitReason ->
             exit(ExitReason)
     end.
+
+add_generators(#state{api_pass_state = ApiPassState,id=Id}=State) ->
+    KeyGen = init_generators(ApiPassState, basho_bench_config:get(key_generator), Id, basho_bench_keygen),
+    ValGen = init_generators(ApiPassState, basho_bench_config:get(value_generator), Id, basho_bench_valgen),
+    State#state{keygen=KeyGen, valgen=ValGen}.
+
+%% Not passing state API - expect non-list spec or error during new attempt
+init_generators(false, Config, Id, Module) when is_tuple(Config) ->
+    Module:new(Config, Id);
+%% Passing state API - process list or turn single-spec into [{default,Generator}]
+init_generators(true, Configs, Id, Module) when is_list(Configs) ->
+    [{N, Module:new(S, Id)} || {N,S} <- Configs].
+
+%% Accessor functions
+
+get_keygen(Name, State) ->
+    proplists:get_value(Name, State#state.keygen).
+
+get_valgen(Name, State) ->
+    proplists:get_value(Name, State#state.valgen).
+
+
```

Comment:
Extra newline here. We want to have a newline at the end of files, but don't need two.